readthedocs/projects/views/public.py

```diff
@@ -117,23 +117,30 @@ def project_badge(request, project_slug):
         style = "flat"
     badge_path = 'projects/badges/%s-' + style + '.svg'
     version_slug = request.GET.get('version', LATEST)
+    version = None
+    url = badge_path % 'unknown'
+
     try:
         version = Version.objects.public(request.user).get(
             project__slug=project_slug, slug=version_slug)
     except Version.DoesNotExist:
-        url = static(badge_path % 'unknown')
-        return HttpResponseRedirect(url)
-    version_builds = version.builds.filter(type='html',
-                                           state='finished').order_by('-date')
-    if not version_builds.exists():
-        url = static(badge_path % 'unknown')
-        return HttpResponseRedirect(url)
-    last_build = version_builds[0]
-    if last_build.success:
-        url = static(badge_path % 'passing')
-    else:
-        url = static(badge_path % 'failing')
-    return HttpResponseRedirect(url)
+        pass
+
+    if version:
+        version_builds = version.builds.filter(type='html',
+                                               state='finished').order_by('-date')
+        if version_builds.exists():
+            last_build = version_builds[0]
+            if last_build.success:
+                url = badge_path % 'passing'
+            else:
+                url = badge_path % 'failing'
+
+    with staticfiles_storage.open(url) as fd:
```

Comment:
Is there a way to put a timeout on this call? We used to do this exact same thing with `shields.io`, but whenever they had downtime, it took down our entire web stack because we'd end up just spinning our web servers looking for badges.

Reply:
We weren't doing 302s to shields.io? I knew we used to use shields.io but I didn't look into precisely how it was implemented. Doing redirects seems like it wouldn't rely on them being up. Badges would be down when shields.io is down but it wouldn't affect our servers at all.

Reply:
It looks like #711 introduced shields.io. It introduced it by streaming the HTTP response from shields which is definitely brittle as you describe. When they're down, it would cause problems for us unless we specify timeouts correctly.

However, it looks like things switched to redirecting to shields.io (in this commit https://github.com/rtfd/readthedocs.org/commit/47757de59b3cad318f13d94ace49aa64f33a36d0 which doesn't really have a PR). That seems like a step in the right direction for sure. #3057 swapped shields.io for self-hosted SVGs but there's no information in the PR as to why that was desirable. To me it seems like a step back.

Reply:
> Is there a way to put a timeout on this call?

To answer this question directly, I don't see a way to add a specific timeout to this call using the version of the Azure libraries we are using. The default timeout is long too (65 seconds).

Reply:
I believe shields had a decent amount of downtime.

Reply:
Looks like this is the issue ticket about it: https://github.com/rtfd/readthedocs.org/issues/2065